File: cpp/D3D12/Minimal-D3D12/main.cpp
#pragma comment(lib, "d3d12")
#pragma comment(lib, "dxgi")
#pragma comment(lib, "d3dcompiler")

// ----------------------------------------------------------------------------------------------------

#include <windows.h>
#include <wrl.h> // ComPtr
#include <d3d12.h>
#include <dxgi1_4.h>
#include <d3dcompiler.h>

#include <memory> // std::memcpy, std::size
#include <cstdlib> // std::exit

// ----------------------------------------------------------------------------------------------------

using Microsoft::WRL::ComPtr;

static const wchar_t* TITLE{ L"Minimal D3D12 by Secticide" };

struct float3 { float x, y, z; };

void check(HRESULT hr) {

    if (FAILED(hr)) {
        std::exit(hr);
    }
}

// ----------------------------------------------------------------------------------------------------

int APIENTRY wWinMain(_In_ HINSTANCE hInstance, _In_opt_ HINSTANCE hPrevInstance, _In_ LPWSTR lpCmdLine, _In_ int nCmdShow) {

    WNDCLASSW wnd_class{ 0, DefWindowProcW, 0, 0, 0, 0, 0, 0, 0, TITLE };

    RegisterClassW(&wnd_class);

    HWND hwnd{ CreateWindowExW(0, TITLE, TITLE, WS_POPUP | WS_MAXIMIZE | WS_VISIBLE, 0, 0, 0, 0, nullptr, nullptr, nullptr, nullptr) };

    // ----------------------------------------------------------------------------------------------------

    ComPtr<ID3D12Debug> debug_controller{};
    check(D3D12GetDebugInterface(IID_PPV_ARGS(&debug_controller)));

    debug_controller->EnableDebugLayer();

    // ----------------------------------------------------------------------------------------------------

    ComPtr<ID3D12Device> device{};
    check(D3D12CreateDevice(nullptr, D3D_FEATURE_LEVEL_12_0, IID_PPV_ARGS(&device)));

    // ----------------------------------------------------------------------------------------------------

    D3D12_COMMAND_QUEUE_DESC queue_desc{};
    queue_desc.Flags = D3D12_COMMAND_QUEUE_FLAG_NONE;
    queue_desc.Type = D3D12_COMMAND_LIST_TYPE_DIRECT;

    ComPtr<ID3D12CommandQueue> cmd_queue{};
    check(device->CreateCommandQueue(&queue_desc, IID_PPV_ARGS(&cmd_queue)));

    // ----------------------------------------------------------------------------------------------------

    ComPtr<IDXGIFactory4> factory{};
    check(CreateDXGIFactory2(DXGI_CREATE_FACTORY_DEBUG, IID_PPV_ARGS(&factory)));

    // ----------------------------------------------------------------------------------------------------

    DXGI_SWAP_CHAIN_DESC1 swapchain_desc{};
    swapchain_desc.BufferCount = 2;
    swapchain_desc.Width = 0; // use window width
    swapchain_desc.Height = 0; // use window height
    swapchain_desc.Format = DXGI_FORMAT_R8G8B8A8_UNORM;
    swapchain_desc.BufferUsage = DXGI_USAGE_RENDER_TARGET_OUTPUT;
    swapchain_desc.SwapEffect = DXGI_SWAP_EFFECT_FLIP_DISCARD;
    swapchain_desc.SampleDesc.Count = 1;

    ComPtr<IDXGISwapChain1> swapchain1{};
    check(factory->CreateSwapChainForHwnd(cmd_queue.Get(), hwnd, &swapchain_desc, nullptr, nullptr, &swapchain1));

    factory->MakeWindowAssociation(hwnd, DXGI_MWA_NO_ALT_ENTER);

    // `IDXGISwapChain3` needed for the `GetCurrentBackBufferIndex` method
    ComPtr<IDXGISwapChain3> swapchain{};
    check(swapchain1->QueryInterface(IID_PPV_ARGS(&swapchain)));

    swapchain->GetDesc1(&swapchain_desc);

    // ----------------------------------------------------------------------------------------------------

    D3D12_DESCRIPTOR_HEAP_DESC rtv_heap_desc{};
    rtv_heap_desc.NumDescriptors = 2;
    rtv_heap_desc.Type = D3D12_DESCRIPTOR_HEAP_TYPE_RTV;

    ComPtr<ID3D12DescriptorHeap> rtv_heap{};
    check(device->CreateDescriptorHeap(&rtv_heap_desc, IID_PPV_ARGS(&rtv_heap)));

    // ----------------------------------------------------------------------------------------------------

    ComPtr<ID3D12Resource> render_targets[2]{};

    UINT rtv_descriptor_size{ device->GetDescriptorHandleIncrementSize(D3D12_DESCRIPTOR_HEAP_TYPE_RTV) };

    D3D12_CPU_DESCRIPTOR_HANDLE rtv_handle{ rtv_heap->GetCPUDescriptorHandleForHeapStart() };

    for (UINT i = 0; i < 2; ++i) {
        check(swapchain->GetBuffer(i, IID_PPV_ARGS(&render_targets[i])));
        device->CreateRenderTargetView(render_targets[i].Get(), nullptr, rtv_handle);
        rtv_handle.ptr = SIZE_T(INT64(rtv_handle.ptr) + INT64(rtv_descriptor_size));
    }

    // ----------------------------------------------------------------------------------------------------

    ComPtr<ID3D12CommandAllocator> cmd_allocator{};
    check(device->CreateCommandAllocator(D3D12_COMMAND_LIST_TYPE_DIRECT, IID_PPV_ARGS(&cmd_allocator)));

    // ----------------------------------------------------------------------------------------------------

    D3D12_ROOT_SIGNATURE_DESC root_signature_desc{};
    root_signature_desc.Flags = D3D12_ROOT_SIGNATURE_FLAG_ALLOW_INPUT_ASSEMBLER_INPUT_LAYOUT;

    ComPtr<ID3DBlob> signature{};
    check(D3D12SerializeRootSignature(&root_signature_desc, D3D_ROOT_SIGNATURE_VERSION_1, &signature, nullptr));

    ComPtr<ID3D12RootSignature> root_signature{};
    check(device->CreateRootSignature(0, signature->GetBufferPointer(), signature->GetBufferSize(), IID_PPV_ARGS(&root_signature)));

    // ----------------------------------------------------------------------------------------------------

    ComPtr<ID3DBlob> vertex_shader{};
    check(D3DCompileFromFile(L"passthrough.hlsl", nullptr, nullptr, "vert_main", "vs_5_0", 0, 0, &vertex_shader, nullptr));

    ComPtr<ID3DBlob> fragment_shader{};
    check(D3DCompileFromFile(L"passthrough.hlsl", nullptr, nullptr, "frag_main", "ps_5_0", 0, 0, &fragment_shader, nullptr));

    // ----------------------------------------------------------------------------------------------------

    D3D12_INPUT_ELEMENT_DESC input_element_descs[]{
        { "POSITION", 0, DXGI_FORMAT_R32G32B32_FLOAT, 0, 0, D3D12_INPUT_CLASSIFICATION_PER_VERTEX_DATA, 0 },
    };

    D3D12_RASTERIZER_DESC rasterizer_desc{};
    rasterizer_desc.FillMode = D3D12_FILL_MODE_SOLID;
    rasterizer_desc.CullMode = D3D12_CULL_MODE_BACK;
    rasterizer_desc.FrontCounterClockwise = FALSE;
    rasterizer_desc.DepthBias = D3D12_DEFAULT_DEPTH_BIAS;
    rasterizer_desc.DepthBiasClamp = D3D12_DEFAULT_DEPTH_BIAS_CLAMP;
    rasterizer_desc.SlopeScaledDepthBias = D3D12_DEFAULT_SLOPE_SCALED_DEPTH_BIAS;
    rasterizer_desc.DepthClipEnable = TRUE;
    rasterizer_desc.MultisampleEnable = FALSE;
    rasterizer_desc.AntialiasedLineEnable = FALSE;
    rasterizer_desc.ForcedSampleCount = 0;
    rasterizer_desc.ConservativeRaster = D3D12_CONSERVATIVE_RASTERIZATION_MODE_OFF;

    D3D12_BLEND_DESC blend_desc{};
    for (UINT i = 0; i < D3D12_SIMULTANEOUS_RENDER_TARGET_COUNT; ++i) {
        D3D12_RENDER_TARGET_BLEND_DESC& rt_blend_desc = blend_desc.RenderTarget[i];
        rt_blend_desc.SrcBlend = D3D12_BLEND_ONE;
        rt_blend_desc.DestBlend = D3D12_BLEND_ZERO;
        rt_blend_desc.BlendOp = D3D12_BLEND_OP_ADD;
        rt_blend_desc.SrcBlendAlpha = D3D12_BLEND_ONE;
        rt_blend_desc.DestBlendAlpha = D3D12_BLEND_ZERO;
        rt_blend_desc.BlendOpAlpha = D3D12_BLEND_OP_ADD;
        rt_blend_desc.LogicOp = D3D12_LOGIC_OP_NOOP;
        rt_blend_desc.RenderTargetWriteMask = D3D12_COLOR_WRITE_ENABLE_ALL;
    }

    D3D12_GRAPHICS_PIPELINE_STATE_DESC pipeline_state_desc{};
    pipeline_state_desc.InputLayout = { input_element_descs, _countof(input_element_descs) };
    pipeline_state_desc.pRootSignature = root_signature.Get();
    pipeline_state_desc.VS = { vertex_shader->GetBufferPointer(), vertex_shader->GetBufferSize() };
    pipeline_state_desc.PS = { fragment_shader->GetBufferPointer(), fragment_shader ->GetBufferSize() };
    pipeline_state_desc.RasterizerState = rasterizer_desc;
    pipeline_state_desc.BlendState = blend_desc;
    pipeline_state_desc.DepthStencilState.DepthEnable = FALSE;
    pipeline_state_desc.DepthStencilState.StencilEnable = FALSE;
    pipeline_state_desc.SampleMask = UINT_MAX;
    pipeline_state_desc.PrimitiveTopologyType = D3D12_PRIMITIVE_TOPOLOGY_TYPE_TRIANGLE;
    pipeline_state_desc.NumRenderTargets = 1;
    pipeline_state_desc.RTVFormats[0] = DXGI_FORMAT_R8G8B8A8_UNORM;
    pipeline_state_desc.SampleDesc.Count = 1;

    ComPtr<ID3D12PipelineState> pipeline_state{};
    check(device->CreateGraphicsPipelineState(&pipeline_state_desc, IID_PPV_ARGS(&pipeline_state)));

    // ----------------------------------------------------------------------------------------------------

    ComPtr<ID3D12GraphicsCommandList> cmd_list{};
    check(device->CreateCommandList(0, D3D12_COMMAND_LIST_TYPE_DIRECT, cmd_allocator.Get(), pipeline_state.Get(), IID_PPV_ARGS(&cmd_list)));

    cmd_list->Close();

    // ----------------------------------------------------------------------------------------------------

    ComPtr<ID3D12Fence> fence{};
    check(device->CreateFence(0, D3D12_FENCE_FLAG_NONE, IID_PPV_ARGS(&fence)));

    UINT64 fence_value{ 1 };
    HANDLE fence_event{ CreateEventW(nullptr, FALSE, FALSE, nullptr) };

    if (fence_event == nullptr) {
        std::exit(GetLastError());
    }

    // ----------------------------------------------------------------------------------------------------

    float3 vertices[]{
        {  0.0f,  0.5f, 0.0f },
        {  0.5f, -0.5f, 0.0f },
        { -0.5f, -0.5f, 0.0f }
    };

    D3D12_HEAP_PROPERTIES heap_props{};
    heap_props.Type = D3D12_HEAP_TYPE_UPLOAD;

    D3D12_RESOURCE_DESC buffer_desc{};
    buffer_desc.Dimension = D3D12_RESOURCE_DIMENSION_BUFFER;
    buffer_desc.Width = sizeof(float3) * std::size(vertices);
    buffer_desc.Height = 1;
    buffer_desc.DepthOrArraySize = 1;
    buffer_desc.MipLevels = 1;
    buffer_desc.SampleDesc.Count = 1;
    buffer_desc.Layout = D3D12_TEXTURE_LAYOUT_ROW_MAJOR;

    ComPtr<ID3D12Resource> vertex_buffer{};
    check(device->CreateCommittedResource(&heap_props, D3D12_HEAP_FLAG_NONE, &buffer_desc, D3D12_RESOURCE_STATE_GENERIC_READ, nullptr, IID_PPV_ARGS(&vertex_buffer)));

    // ----------------------------------------------------------------------------------------------------

    void* vertex_data_begin{};
    D3D12_RANGE read_range{ 0, 0 };

    check(vertex_buffer->Map(0, &read_range, &vertex_data_begin));

    std::memcpy(vertex_data_begin, std::data(vertices), sizeof(float3) * std::size(vertices));

    vertex_buffer->Unmap(0, nullptr);

    // ----------------------------------------------------------------------------------------------------

    D3D12_VERTEX_BUFFER_VIEW vertex_buffer_view{};
    vertex_buffer_view.BufferLocation = vertex_buffer->GetGPUVirtualAddress();
    vertex_buffer_view.StrideInBytes = sizeof(float3);
    vertex_buffer_view.SizeInBytes = sizeof(float3) * std::size(vertices);

    D3D12_VIEWPORT viewport{};
    viewport.Width = (float)swapchain_desc.Width;
    viewport.Height = (float)swapchain_desc.Height;
    viewport.MinDepth = D3D12_MIN_DEPTH;
    viewport.MaxDepth = D3D12_MAX_DEPTH;

    D3D12_RECT scissor{};
    scissor.right = swapchain_desc.Width;
    scissor.bottom = swapchain_desc.Height;

    // ----------------------------------------------------------------------------------------------------

    UINT frame_index{ swapchain->GetCurrentBackBufferIndex() };
    bool is_running{ true };

    while (is_running) {

        MSG msg{};
        while (PeekMessageW(&msg, nullptr, 0, 0, PM_REMOVE)) {

            if (msg.message == WM_KEYDOWN) {
                is_running = false;
            }

            TranslateMessage(&msg);
            DispatchMessage(&msg);
        }

        cmd_allocator->Reset();

        cmd_list->Reset(cmd_allocator.Get(), pipeline_state.Get());
        cmd_list->SetGraphicsRootSignature(root_signature.Get());
        cmd_list->RSSetViewports(1, &viewport);
        cmd_list->RSSetScissorRects(1, &scissor);

        D3D12_RESOURCE_BARRIER barrier{};
        barrier.Type = D3D12_RESOURCE_BARRIER_TYPE_TRANSITION;
        barrier.Flags = D3D12_RESOURCE_BARRIER_FLAG_NONE;
        barrier.Transition = {
            render_targets[frame_index].Get(),
            D3D12_RESOURCE_BARRIER_ALL_SUBRESOURCES,
            D3D12_RESOURCE_STATE_PRESENT,
            D3D12_RESOURCE_STATE_RENDER_TARGET
        };
        cmd_list->ResourceBarrier(1, &barrier);

        rtv_handle = rtv_heap->GetCPUDescriptorHandleForHeapStart();
        rtv_handle.ptr = SIZE_T(INT64(rtv_handle.ptr) + INT64(frame_index) * INT64(rtv_descriptor_size));
        cmd_list->OMSetRenderTargets(1, &rtv_handle, FALSE, nullptr);

        const float clear_colour[]{ 0.0f, 0.2f, 0.4f, 1.0f };
        cmd_list->ClearRenderTargetView(rtv_handle, clear_colour, 0, nullptr);
        cmd_list->IASetPrimitiveTopology(D3D_PRIMITIVE_TOPOLOGY_TRIANGLELIST);
        cmd_list->IASetVertexBuffers(0, 1, &vertex_buffer_view);
        cmd_list->DrawInstanced(3, 1, 0, 0);

        barrier.Transition = {
            render_targets[frame_index].Get(),
            D3D12_RESOURCE_BARRIER_ALL_SUBRESOURCES,
            D3D12_RESOURCE_STATE_RENDER_TARGET,
            D3D12_RESOURCE_STATE_PRESENT
        };
        cmd_list->ResourceBarrier(1, &barrier);
        cmd_list->Close();

        ID3D12CommandList* cmd_lists[]{ cmd_list.Get() };
        cmd_queue->ExecuteCommandLists(std::size(cmd_lists), cmd_lists);

        swapchain->Present(1, 0);

        const UINT64 current_fence_value{ fence_value };
        cmd_queue->Signal(fence.Get(), current_fence_value);
        fence_value += 1;

        if (fence->GetCompletedValue() < current_fence_value) {
            fence->SetEventOnCompletion(current_fence_value, fence_event);
            WaitForSingleObject(fence_event, INFINITE);
        }

        frame_index = swapchain->GetCurrentBackBufferIndex();
    }

    ShowWindow(hwnd, SW_HIDE);
    CloseHandle(fence_event);

    return 0;
}

File: rust/D3D12/minimal-d3d12/src/main.rs
use windows::{
    core::*,
    Win32::{
        Foundation::*,
        Graphics::{
            Direct3D, Direct3D12,
            Dxgi::{self, Common},
        },
        System::Threading,
        UI::WindowsAndMessaging::*,
    },
};

use std::{mem::ManuallyDrop, ptr};

const TITLE: PCWSTR = windows::core::w!("Minimal D3D12 by Secticide");
const PASSTHROUGH: &'static std::ffi::CStr = c"
struct vertex_output
{
    float4 position : SV_POSITION;
};

vertex_output vert_main(float3 position : POSITION)
{
    vertex_output output;
    output.position = float4(position, 1.0f);
    return output;
}

float4 frag_main(vertex_output input) : SV_TARGET
{
    return float4(1.0f, 0.0f, 0.0f, 1.0f);
}";

#[allow(dead_code)]
struct Float3 {
    x: f32,
    y: f32,
    z: f32,
}

extern "system" fn window_proc(hwnd: HWND, msg: u32, wparam: WPARAM, lparam: LPARAM) -> LRESULT {
    unsafe { DefWindowProcW(hwnd, msg, wparam, lparam) }
}

// ----------------------------------------------------------------------------------------------------

fn main() -> Result<()> {
    unsafe {
        let wnd_class = WNDCLASSW {
            lpfnWndProc: Some(window_proc),
            lpszClassName: TITLE,
            ..Default::default()
        };

        RegisterClassW(&wnd_class);

        let hwnd = CreateWindowExW(
            WINDOW_EX_STYLE(0),
            TITLE,
            TITLE,
            WS_POPUP | WS_MAXIMIZE | WS_VISIBLE,
            0,
            0,
            0,
            0,
            None,
            None,
            None,
            None,
        )
        .unwrap();

        // ----------------------------------------------------------------------------------------------------

        let debug_controller: Direct3D12::ID3D12Debug = {
            let mut result = None;
            Direct3D12::D3D12GetDebugInterface(&mut result)?;
            result.unwrap()
        };

        debug_controller.EnableDebugLayer();

        // ----------------------------------------------------------------------------------------------------

        let device: Direct3D12::ID3D12Device = {
            let mut result = None;
            Direct3D12::D3D12CreateDevice(None, Direct3D::D3D_FEATURE_LEVEL_12_0, &mut result)?;
            result.unwrap_unchecked()
        };

        // ----------------------------------------------------------------------------------------------------

        let queue_desc = Direct3D12::D3D12_COMMAND_QUEUE_DESC {
            Type: Direct3D12::D3D12_COMMAND_LIST_TYPE_DIRECT,
            Flags: Direct3D12::D3D12_COMMAND_QUEUE_FLAG_NONE,
            ..Default::default()
        };

        let cmd_queue: Direct3D12::ID3D12CommandQueue = device.CreateCommandQueue(&queue_desc)?;

        // ----------------------------------------------------------------------------------------------------

        let factory: Dxgi::IDXGIFactory4 = Dxgi::CreateDXGIFactory2(Dxgi::DXGI_CREATE_FACTORY_DEBUG)?;

        // ----------------------------------------------------------------------------------------------------

        let swapchain_desc = Dxgi::DXGI_SWAP_CHAIN_DESC1 {
            Format: Common::DXGI_FORMAT_R8G8B8A8_UNORM,
            SampleDesc: Common::DXGI_SAMPLE_DESC { Count: 1, Quality: 0, },
            BufferUsage: Dxgi::DXGI_USAGE_RENDER_TARGET_OUTPUT,
            BufferCount: 2,
            SwapEffect: Dxgi::DXGI_SWAP_EFFECT_FLIP_DISCARD,
            ..Default::default()
        };

        let swapchain: Dxgi::IDXGISwapChain1 =
            factory.CreateSwapChainForHwnd(&cmd_queue, hwnd, &swapchain_desc, None, None)?;
        factory.MakeWindowAssociation(hwnd, Dxgi::DXGI_MWA_NO_ALT_ENTER)?;

        // `IDXGISwapChain3` needed for the `GetCurrentBackBufferIndex` method
        let swapchain: Dxgi::IDXGISwapChain3 = swapchain.cast()?;
        let swapchain_desc = swapchain.GetDesc1()?;

        // ----------------------------------------------------------------------------------------------------

        let rtv_descriptor_heap = Direct3D12::D3D12_DESCRIPTOR_HEAP_DESC {
            Type: Direct3D12::D3D12_DESCRIPTOR_HEAP_TYPE_RTV,
            NumDescriptors: 2,
            ..Default::default()
        };

        let rtv_heap: Direct3D12::ID3D12DescriptorHeap =
            device.CreateDescriptorHeap(&rtv_descriptor_heap)?;

        // ----------------------------------------------------------------------------------------------------

        let rtv_descriptor_size =
            device.GetDescriptorHandleIncrementSize(Direct3D12::D3D12_DESCRIPTOR_HEAP_TYPE_RTV);
        let mut rtv_handle = rtv_heap.GetCPUDescriptorHandleForHeapStart();

        let render_targets: [Direct3D12::ID3D12Resource; 2] =
            [swapchain.GetBuffer(0)?, swapchain.GetBuffer(1)?];

        for target in render_targets.iter() {
            device.CreateRenderTargetView(target, None, rtv_handle);
            rtv_handle.ptr += rtv_descriptor_size as usize;
        }

        // ----------------------------------------------------------------------------------------------------

        let cmd_allocator: Direct3D12::ID3D12CommandAllocator =
            device.CreateCommandAllocator(Direct3D12::D3D12_COMMAND_LIST_TYPE_DIRECT)?;

        // ----------------------------------------------------------------------------------------------------

        let root_signature_desc = Direct3D12::D3D12_ROOT_SIGNATURE_DESC {
            Flags: Direct3D12::D3D12_ROOT_SIGNATURE_FLAG_ALLOW_INPUT_ASSEMBLER_INPUT_LAYOUT,
            ..Default::default()
        };

        let signature = {
            let mut result = None;
            Direct3D12::D3D12SerializeRootSignature(
                &root_signature_desc,
                Direct3D12::D3D_ROOT_SIGNATURE_VERSION_1,
                &mut result,
                None,
            )?;
            result.unwrap_unchecked()
        };

        let signature_slice = std::slice::from_raw_parts(
            signature.GetBufferPointer() as *const _,
            signature.GetBufferSize(),
        );
        let root_signature: Direct3D12::ID3D12RootSignature =
            device.CreateRootSignature(0, signature_slice)?;

        // ----------------------------------------------------------------------------------------------------

        let vertex_shader = {
            let mut result = None;
            Direct3D::Fxc::D3DCompile(
                PASSTHROUGH.as_ptr() as _,
                PASSTHROUGH.count_bytes(),
                PSTR(ptr::null_mut()),
                None,
                None,
                s!("vert_main"),
                s!("vs_5_0"),
                0,
                0,
                &mut result,
                None,
            )?;
            result.unwrap_unchecked()
        };

        let fragment_shader = {
            let mut result = None;
            Direct3D::Fxc::D3DCompile(
                PASSTHROUGH.as_ptr() as _,
                PASSTHROUGH.count_bytes(),
                PSTR(ptr::null_mut()),
                None,
                None,
                s!("frag_main"),
                s!("ps_5_0"),
                0,
                0,
                &mut result,
                None,
            )?;
            result.unwrap_unchecked()
        };

        // ----------------------------------------------------------------------------------------------------

        let input_element_descs = [Direct3D12::D3D12_INPUT_ELEMENT_DESC {
            SemanticName: s!("POSITION"),
            SemanticIndex: 0,
            Format: Common::DXGI_FORMAT_R32G32B32_FLOAT,
            InputSlot: 0,
            AlignedByteOffset: 0,
            InputSlotClass: Direct3D12::D3D12_INPUT_CLASSIFICATION_PER_VERTEX_DATA,
            InstanceDataStepRate: 0,
        }];

        let rasterizer_desc = Direct3D12::D3D12_RASTERIZER_DESC {
            FillMode: Direct3D12::D3D12_FILL_MODE_SOLID,
            CullMode: Direct3D12::D3D12_CULL_MODE_BACK,
            FrontCounterClockwise: false.into(),
            DepthBias: Direct3D12::D3D12_DEFAULT_DEPTH_BIAS,
            DepthBiasClamp: Direct3D12::D3D12_DEFAULT_DEPTH_BIAS_CLAMP,
            SlopeScaledDepthBias: Direct3D12::D3D12_DEFAULT_SLOPE_SCALED_DEPTH_BIAS,
            DepthClipEnable: true.into(),
            MultisampleEnable: false.into(),
            AntialiasedLineEnable: false.into(),
            ForcedSampleCount: 0,
            ConservativeRaster: Direct3D12::D3D12_CONSERVATIVE_RASTERIZATION_MODE_OFF,
        };

        let blend_desc = {
            let mut result = Direct3D12::D3D12_BLEND_DESC::default();
            for rt in result.RenderTarget.iter_mut() {
                rt.SrcBlend = Direct3D12::D3D12_BLEND_ONE;
                rt.DestBlend = Direct3D12::D3D12_BLEND_ZERO;
                rt.BlendOp = Direct3D12::D3D12_BLEND_OP_ADD;
                rt.SrcBlendAlpha = Direct3D12::D3D12_BLEND_ONE;
                rt.DestBlendAlpha = Direct3D12::D3D12_BLEND_ZERO;
                rt.BlendOpAlpha = Direct3D12::D3D12_BLEND_OP_ADD;
                rt.LogicOp = Direct3D12::D3D12_LOGIC_OP_NOOP;
                rt.RenderTargetWriteMask = Direct3D12::D3D12_COLOR_WRITE_ENABLE_ALL.0 as u8;
            }
            result
        };

        let mut pipeline_state_desc = Direct3D12::D3D12_GRAPHICS_PIPELINE_STATE_DESC {
            pRootSignature: ManuallyDrop::new(Some(std::mem::transmute_copy(&root_signature))),
            VS: Direct3D12::D3D12_SHADER_BYTECODE {
                pShaderBytecode: vertex_shader.GetBufferPointer(),
                BytecodeLength: vertex_shader.GetBufferSize(),
            },
            PS: Direct3D12::D3D12_SHADER_BYTECODE {
                pShaderBytecode: fragment_shader.GetBufferPointer(),
                BytecodeLength: fragment_shader.GetBufferSize(),
            },
            BlendState: blend_desc,
            SampleMask: std::u32::MAX,
            RasterizerState: rasterizer_desc,
            DepthStencilState: Direct3D12::D3D12_DEPTH_STENCIL_DESC {
                DepthEnable: false.into(),
                StencilEnable: false.into(),
                ..Default::default()
            },
            InputLayout: Direct3D12::D3D12_INPUT_LAYOUT_DESC {
                pInputElementDescs: input_element_descs.as_ptr(),
                NumElements: input_element_descs.len() as u32,
            },
            PrimitiveTopologyType: Direct3D12::D3D12_PRIMITIVE_TOPOLOGY_TYPE_TRIANGLE,
            NumRenderTargets: 1,
            SampleDesc: Common::DXGI_SAMPLE_DESC {
                Count: 1,
                Quality: 0,
            },
            ..Default::default()
        };

        // Only set the first format
        pipeline_state_desc.RTVFormats[0] = Common::DXGI_FORMAT_R8G8B8A8_UNORM;

        let pipeline_state: Direct3D12::ID3D12PipelineState =
            device.CreateGraphicsPipelineState(&pipeline_state_desc)?;

        // ----------------------------------------------------------------------------------------------------

        let cmd_list: Direct3D12::ID3D12GraphicsCommandList = device.CreateCommandList(
            0,
            Direct3D12::D3D12_COMMAND_LIST_TYPE_DIRECT,
            &cmd_allocator,
            &pipeline_state,
        )?;

        let _ = cmd_list.Close();

        // ----------------------------------------------------------------------------------------------------

        let fence: Direct3D12::ID3D12Fence =
            device.CreateFence(0, Direct3D12::D3D12_FENCE_FLAG_NONE)?;

        let mut fence_value = 1;
        let fence_event = Threading::CreateEventW(None, false, false, None)?;

        // ----------------------------------------------------------------------------------------------------

        let vertices = [
            Float3 { x: 0.0, y: 0.5, z: 0.0, },
            Float3 { x: 0.5, y: -0.5, z: 0.0, },
            Float3 { x: -0.5, y: -0.5, z: 0.0, },
        ];

        let heap_props = Direct3D12::D3D12_HEAP_PROPERTIES {
            Type: Direct3D12::D3D12_HEAP_TYPE_UPLOAD,
            ..Default::default()
        };

        let buffer_desc = Direct3D12::D3D12_RESOURCE_DESC {
            Dimension: Direct3D12::D3D12_RESOURCE_DIMENSION_BUFFER,
            Width: std::mem::size_of_val(&vertices) as u64,
            Height: 1,
            DepthOrArraySize: 1,
            MipLevels: 1,
            SampleDesc: Common::DXGI_SAMPLE_DESC {
                Count: 1,
                Quality: 0,
            },
            Layout: Direct3D12::D3D12_TEXTURE_LAYOUT_ROW_MAJOR,
            ..Default::default()
        };

        let vertex_buffer: Direct3D12::ID3D12Resource = {
            let mut result = None;
            device.CreateCommittedResource(
                &heap_props,
                Direct3D12::D3D12_HEAP_FLAG_NONE,
                &buffer_desc,
                Direct3D12::D3D12_RESOURCE_STATE_GENERIC_READ,
                None,
                &mut result,
            )?;
            result.unwrap_unchecked()
        };

        // ----------------------------------------------------------------------------------------------------

        let mut vertex_data_begin = ptr::null_mut();
        let read_range = Direct3D12::D3D12_RANGE { Begin: 0, End: 0 };

        vertex_buffer.Map(0, Some(&read_range), Some(&mut vertex_data_begin))?;

        ptr::copy_nonoverlapping::<Float3>(vertices.as_ptr(), vertex_data_begin as _, vertices.len());

        vertex_buffer.Unmap(0, None);

        // ----------------------------------------------------------------------------------------------------

        let vertex_buffer_view = Direct3D12::D3D12_VERTEX_BUFFER_VIEW {
            BufferLocation: vertex_buffer.GetGPUVirtualAddress(),
            SizeInBytes: std::mem::size_of_val(&vertices) as u32,
            StrideInBytes: std::mem::size_of::<Float3>() as u32,
        };

        let viewport = Direct3D12::D3D12_VIEWPORT {
            Width: swapchain_desc.Width as f32,
            Height: swapchain_desc.Height as f32,
            MinDepth: Direct3D12::D3D12_MIN_DEPTH,
            MaxDepth: Direct3D12::D3D12_MAX_DEPTH,
            ..Default::default()
        };

        let scissor = RECT {
            right: swapchain_desc.Width as i32,
            bottom: swapchain_desc.Height as i32,
            ..Default::default()
        };

        // ----------------------------------------------------------------------------------------------------

        let mut frame_index = swapchain.GetCurrentBackBufferIndex();
        let mut is_running = true;

        while is_running {
            let mut msg = MSG::default();
            while PeekMessageW(&mut msg, None, 0, 0, PM_REMOVE).into() {
                if msg.message == WM_KEYDOWN {
                    is_running = false;
                }

                let _ = TranslateMessage(&msg);
                DispatchMessageW(&msg);
            }

            cmd_allocator.Reset()?;

            cmd_list.Reset(&cmd_allocator, &pipeline_state)?;
            cmd_list.SetGraphicsRootSignature(&root_signature);
            cmd_list.RSSetViewports(&[viewport]);
            cmd_list.RSSetScissorRects(&[scissor]);

            let barrier = Direct3D12::D3D12_RESOURCE_BARRIER {
                Type: Direct3D12::D3D12_RESOURCE_BARRIER_TYPE_TRANSITION,
                Flags: Direct3D12::D3D12_RESOURCE_BARRIER_FLAG_NONE,
                Anonymous: Direct3D12::D3D12_RESOURCE_BARRIER_0 {
                    Transition: ManuallyDrop::new(Direct3D12::D3D12_RESOURCE_TRANSITION_BARRIER {
                        pResource: ManuallyDrop::new(Some(std::mem::transmute_copy(
                            &render_targets[frame_index as usize],
                        ))),
                        Subresource: Direct3D12::D3D12_RESOURCE_BARRIER_ALL_SUBRESOURCES,
                        StateBefore: Direct3D12::D3D12_RESOURCE_STATE_PRESENT,
                        StateAfter: Direct3D12::D3D12_RESOURCE_STATE_RENDER_TARGET,
                    }),
                },
            };
            cmd_list.ResourceBarrier(&[barrier]);

            rtv_handle = rtv_heap.GetCPUDescriptorHandleForHeapStart();
            rtv_handle.ptr += frame_index as usize * rtv_descriptor_size as usize;
            cmd_list.OMSetRenderTargets(1, Some(&rtv_handle), false, None);

            const CLEAR_COLOUR: [f32; 4] = [0.0, 0.2, 0.4, 1.0];
            cmd_list.ClearRenderTargetView(rtv_handle, &CLEAR_COLOUR, None);
            cmd_list.IASetPrimitiveTopology(Direct3D::D3D_PRIMITIVE_TOPOLOGY_TRIANGLELIST);
            cmd_list.IASetVertexBuffers(0, Some(&[vertex_buffer_view]));
            cmd_list.DrawInstanced(3, 1, 0, 0);

            let barrier = Direct3D12::D3D12_RESOURCE_BARRIER {
                Type: Direct3D12::D3D12_RESOURCE_BARRIER_TYPE_TRANSITION,
                Flags: Direct3D12::D3D12_RESOURCE_BARRIER_FLAG_NONE,
                Anonymous: Direct3D12::D3D12_RESOURCE_BARRIER_0 {
                    Transition: ManuallyDrop::new(Direct3D12::D3D12_RESOURCE_TRANSITION_BARRIER {
                        pResource: ManuallyDrop::new(Some(std::mem::transmute_copy(
                            &render_targets[frame_index as usize],
                        ))),
                        Subresource: Direct3D12::D3D12_RESOURCE_BARRIER_ALL_SUBRESOURCES,
                        StateBefore: Direct3D12::D3D12_RESOURCE_STATE_RENDER_TARGET,
                        StateAfter: Direct3D12::D3D12_RESOURCE_STATE_PRESENT,
                    }),
                },
            };
            cmd_list.ResourceBarrier(&[barrier]);
            cmd_list.Close()?;

            cmd_queue.ExecuteCommandLists(&[Some(cmd_list.cast()?)]);

            swapchain.Present(1, Dxgi::DXGI_PRESENT(0)).ok()?;

            let current_fence_value = fence_value;
            cmd_queue.Signal(&fence, current_fence_value)?;
            fence_value += 1;

            if fence.GetCompletedValue() < current_fence_value {
                fence.SetEventOnCompletion(current_fence_value, fence_event)?;
                Threading::WaitForSingleObject(fence_event, Threading::INFINITE);
            }

            frame_index = swapchain.GetCurrentBackBufferIndex();
        }

        let _ = ShowWindow(hwnd, SW_HIDE);
        let _ = CloseHandle(fence_event);

        Ok(())
    }
}
